### DISCUSSION TITLE
Software design book/course advice

### DISCUSSION THREAD

**Discussion by biomed12** [techexcellenceio/techexcellence#41] [Q&A]
Hi all,

First of all thanks to people who spend their time to create such a community.

My question is all about traditional way of software design. I am comfortable with TDD. I already shipped a few non-trivial products at middleware and embedded level. However the industry, at least the bubble I live, is not that idealistic. Still I am supposed to do some design-up-front via some UML diagramming tools. I hate this way of working but I have to do that for professional reasons.

So, how can I get comfortable with this way of working? Is there any popular software design books or courses...? What could be the most efficient way to achieve some progress? I think I understand SOLID, I know about some architectural patterns and traditional design patterns etc. Still I don't feel comfortable while doing DUF. I hope I could give the question clearly.

Regards,
BK

**Comment by marciovmartins** [comment]
I would suggest some DDD book to get you into domain and bounded context. This will help you to do better questions to find the use cases. Once the use cases and domains are in place the diagram emerge by itself.

When I worked in this kind of environment, those diagrams did not last long and were abandon after some time. So, I did not invested mutch time doing something great or perfect because I knew this would not be used for so long.
In those situation I invested in some onion architecture to make the code easier to change after some breakthrough that the company have.
